src/Type/TypeCombinator.php

```diff
@@ -461,34 +466,55 @@ private static function compareTypesInUnion(Type $a, Type $b): ?array
 			&& ($a->describe(VerbosityLevel::value()) === 'non-empty-string'
 				|| $a->describe(VerbosityLevel::value()) === 'non-falsy-string')
 		) {
-			return [new StringType(), null];
+			return [self::intersect(
+				new StringType(),
+				...self::getAccessoryCaseStringTypes($a),
+			), null];
 		}
 
 		if (
 			$a instanceof ConstantStringType
 			&& $a->getValue() === '0'
```

Comment:
This looks like we need another test (the change seems untested)

Reply:
Done